### PR TITLE
[PFC] Test if PFC can pause lossless priorities

### PR DESCRIPTION
**Pull request by baiwei0427**
### Description of PR

Signed-off-by: Wei Bai baiwei0427@gmail.com

Summary: Two pytest tests:
- Test if PFC can pause lossless priorities without affecting the other priorities
- Test if lossless and lossy priorities can forward packets in the absence of PFC pause frames

### Type of change
- [] Test case(new/improvement)

#### How did you do it?
1. Set up the testbed, including (1) copy the PFC packet generator toe the leaf fanout switches, and (2) copying the PTF script to the PTF container.

2. Given a tested interface _i_ on the DUT, we generate PFC pause frames at the leaf fanout to _block_ the lossless priority _p_ of this interface.

3. The PTF script transmits two packets which should leave DUT from the interface _i_. One packet is mapped to priority _p_. The other packet is mapped to any other priority.

4. The PTF script should only receive the packet that is not mapped to priority _p_. 

#### How did you verify/test it?
I ran this test case in our testbed.

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
Currently, the test case only supports T0 topology. I will add more topologies in the future.

### CODE REVIEW

File: tests/qos/test_pfc_pause.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+""" Maximum number of interfaces to test on a DUT """
+MAX_TEST_INTFS_COUNT = 4
+         
+def atoi(text):
+    return int(text) if text.isdigit() else text
+
+def natural_keys(text):
+    return [atoi(c) for c in re.split(r'(\d+)', text)]
+
+def setup_testbed(ansible_adhoc, testbed, leaf_fanouts):
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: can you move this function to the helper file? These steps will be applicable to multiple PFC tests

File: tests/qos/test_pfc_pause.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    files = [ansible_stdout_to_str(x['path']) for x in result]
+    
+    for file in files:
+        ptf_ans.file(path=file, mode="absent")
+
+    """ Copy the PFC test script to the PTF container """  
+    file_src = os.path.join(os.path.dirname(__file__), PTF_FILE_RELATIVE_PATH)
+    ptf_ans.copy(src=file_src, dest=PTF_FILE_DEST, force=True)
+    
+def run_test_t0(ansible_adhoc,
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: I would suggest to break down this function into smaller functions. vlan setup should be in its own function, mac population steps should be in another one. start and stop pfc storm should be a separate function These are again part of the setup that needs to be done before the test run. So they should be called by a setup function. The run test should start the pfc storm, run the traffic and verify functionality

File: tests/qos/test_pfc_pause.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    """ Generate IP addresses for servers in the Vlan """
+    vlan_ip_addrs = get_addrs_in_subnet(vlan_subnet, len(vlan_members))
+    """ Generate MAC addresses 00:00:00:00:00:XX for servers in the Vlan """
+    vlan_mac_addrs = [5 * '00:' + format(k, '02x') for k in random.sample(range(1, 256), len(vlan_members))]
+    
+    """ Find correspoinding interfaces on PTF """
+    intf_facts = dut_ans.interface_facts()['ansible_facts']['ansible_interface_facts'] 
+    phy_intfs = [k for k in intf_facts.keys() if k.startswith('Ethernet')]
+    phy_intfs.sort(key = natural_keys)
+    vlan_members.sort(key = natural_keys)
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: avoid spacing between '='

File: tests/qos/test_pfc_pause.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
+
+from ansible_host import AnsibleHost
+from qos_fixtures import *
+from qos_helpers import ansible_stdout_to_str, eos_to_linux_intf, get_active_intfs, get_addrs_in_subnet, get_neigh_ip,check_mac_table, get_mac, config_intf_ip_mac
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: import only the ones that will used in the test

File: tests/qos/test_pfc_pause.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+            cmd = "nohup sudo python %s -i %s -g -t %d -n %d </dev/null >/dev/null 2>&1 &" % (PFC_GEN_FILE_DEST, peer_port_name, pause_time, PFC_PKT_COUNT)
+        
+            if pfc_pause:
+                cmd = "nohup sudo python %s -i %s -p %d -t %d -n %d </dev/null >/dev/null 2>&1 &" % (PFC_GEN_FILE_DEST, peer_port_name, 2**pause_prio, pause_time, PFC_PKT_COUNT)
+                
+            """ Start PFC / FC storm """
+            peerdev_ans.shell(cmd)
+       
+            """ Wait for PFC pause frame generation """
+            time.sleep(2)
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: These should be in a helper function. you are going to need it in multiple tests

File: tests/qos/test_pfc_pause.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+        else:
+            results[dut_intf_paused] = [int(words[1]), int(words[3])] 
+        time.sleep(1)
+
+        if send_pause:
+            """ Stop PFC / FC storm """
+            cmd = "sudo kill -9 $(pgrep -f %s) </dev/null >/dev/null 2>&1 &" % (PFC_GEN_FILE)
+            peerdev_ans.shell(cmd)
+            time.sleep(1)
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: move this to a helper function

File: tests/qos/test_pfc_pause.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+        src_mac = ptf_mac_addrs[src_index]
+        dst_mac = ptf_mac_addrs[dst_index]
+       
+        """ DUT interface to pause """
+        dut_intf_paused = dut_intfs[dst_index]
+                                
+        """ Clear MAC table in DUT (host memory and ASIC) """
+        dut_ans.shell('sonic-clear fdb all </dev/null >/dev/null 2>&1 &')
+        dut_ans.shell('sudo ip -s -s neigh flush all </dev/null >/dev/null 2>&1 &')
+        time.sleep(2)
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: is this still needed? it is done in config_testbed_t0?

File: tests/qos/test_pfc_pause.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+                        continue
+            
+                    if pass_count < total_count * PTF_PASS_RATIO_THRESH:
+                        errors[intf] = results[intf]
+
+                if len(errors) > 0:
+                    print "errors occured:\n{}".format("\n".join(errors))
+                    assert 0 
+
+def test_no_pfc(ansible_adhoc,
```
Comment (neethajohn) [inline tests/qos/test_pfc_pause.py]: Please update the PR description to include this test as well

File: ansible/roles/test/files/ptftests/pfc_pause_test.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+            pkts = capture_matched_packets(self, masked_exp_pkt, self.port_dst)
+            
+            time.sleep(self.pkt_intvl)
+            
+            """ If the queue is paused, we should only receive the background packet """
+            if self.queue_paused:
+                pass_cnt += int(len(pkts) == 1 and scapy.Ether(pkts[0])[scapy.IP].tos == tos_bg)
+
+            else:
+                pass_cnt += int(len(pkts) == 2)
```
[the remark this answers is not in the document]
Comment (baiwei0427) [reply ansible/roles/test/files/ptftests/pfc_pause_test.py]: @neethajohn Sending many packets may take some time. As a result, if we send all packets and then check, we may miss the first several packets.

File: tests/qos/test_pfc_pause.py (hunk 1 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_pause.py]: @neethajohn Fixed.

File: tests/qos/test_pfc_pause.py (hunk 5 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_pause.py]: Fixed.

File: tests/qos/test_pfc_pause.py (hunk 8 above)
Comment (baiwei0427) [reply tests/qos/test_pfc_pause.py]: Fixed

File: tests/qos/test_pfc_pause.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    @param queue_paused: if the queue is expected to be paused
+    @param is_pfc: If this test is for PFC?
+    @param pause_time: pause time quanta of PFC frames. It is 65535 (maximum pause time quanta) by default.
+    @param max_test_count: maximum count of tests to run. By default, it is a very large value to cover all the interfaces.
+    return: Return # of iterations and # of passed iterations for each test case. 
+    """
+    
+    print testbed 
+    if testbed['topo']['name'].startswith('t0'):
+        return run_test_t0(ansible_adhoc=ansible_adhoc,
```
[the remark this answers is not in the document]
Comment (neethajohn) [reply tests/qos/test_pfc_pause.py]: We can come back to this once you run on T1

File: tests/qos/qos_helpers.py [hunk 11]
```diff
@@ -1,7 +1,221 @@
...
 def eos_to_linux_intf(eos_intf_name):
     """
     @Summary: Map EOS's interface name to Linux's interface name
     @param eos_intf_name: Interface name in EOS
     @return: Return the interface name in Linux
     """
-    return eos_intf_name.replace('Ethernet', 'et').replace('/', '_') 
+    return eos_intf_name.replace('Ethernet', 'et').replace('/', '_')
+
+def get_phy_intfs(host_ans):
```
Comment (yxieca) [inline tests/qos/qos_helpers.py]: This function should be moved to tests/common/platform/interface_utils.py

But do it with another PR.
